eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/OnTheFlyFunding.scala

```diff
@@ -38,7 +38,29 @@ import scala.concurrent.duration.FiniteDuration
 
 object OnTheFlyFunding {
 
-  case class Config(proposalTimeout: FiniteDuration)
+  case class Config(proposalTimeout: FiniteDuration) {
+    // When funding a transaction using from_future_htlc, we are taking the risk that the remote node doesn't fulfill
+    // the corresponding HTLCs. If we detect that our peer fails such HTLCs, we automatically disable from_future_htlc
+    // to limit our exposure.
+    // Note that this state is flushed when restarting: node operators should explicitly remove the from_future_htlc
+    // payment type from their liquidity ads configuration if they want to keep it disabled.
+    private val suspectFromFutureHtlcRelays = scala.collection.concurrent.TrieMap.empty[ByteVector32, PublicKey]
+
+    /** We allow using from_future_htlc if we don't have any pending payment that is abusing it. */
+    def isFromFutureHtlcAllowed: Boolean = suspectFromFutureHtlcRelays.isEmpty
+
+    /** Disable from_future_htlc payments because the provided payment_hash seems malicious. */
+    def disableFromFutureHtlc(paymentHash: ByteVector32, remoteNodeId: PublicKey): Unit = suspectFromFutureHtlcRelays.addOne(paymentHash, remoteNodeId)
```

Comment:
I would maintain symmetry between `fromFutureHtlcFailed`/`fromFutureHtlcFulfilled` and reserve the decision to `isFromFutureHtlcAllowed` (where we could potentially use a threshold).

Also we need a `Gauge` metric to track the current number of suspicious relays.

```suggestion
    /** If a from_future_htlc fails after liquidity has been added, it is potentially malicious and we add it to the list of suspicious payment hashes. */
    def fromFutureHtlcFailed(paymentHash: ByteVector32, remoteNodeId: PublicKey): Unit = suspectFromFutureHtlcRelays.addOne(paymentHash, remoteNodeId)
```